### PR TITLE
Remove the use of cTrait.default_value to set the default value

### PR DESCRIPTION
**Pull request by mdickinson**
The use of `cTrait.default_value` to _set_ the default value type and default value was deprecated in 2019, in #620. This PR removes that deprecated support. To set the default value type and default value for a `CTrait` or `cTrait` instance, use `set_default_value`.

**Checklist**
- [x] Tests
- [x] Update API reference (`docs/source/traits_api_reference`)  N/A
- [ ] Update User manual (`docs/source/traits_user_manual`)  N/A
- [ ] Update type annotation hints in `traits-stubs`  N/A

### CODE REVIEW

File: traits/ctraits.c [hunk 1]
```diff
@@ -3078,34 +3078,23 @@ _trait_set_default_value(trait_object *trait, PyObject *args)
 }
 
 /*-----------------------------------------------------------------------------
-|  Get or set the 'default_value_type' and 'default_value' fields
-|  of a CTrait instance. Use of this function for setting the default
-|  value information is deprecated; use set_default_value instead.
+|  Get or the 'default_value_type' and 'default_value' fields
```
Comment (rahulporuri) [inline traits/ctraits.c:3081]:
```suggestion
|  Get the 'default_value_type' and 'default_value' fields
```

File: traits/ctraits.c [hunk 2]
```diff
@@ -3078,34 +3078,23 @@ _trait_set_default_value(trait_object *trait, PyObject *args)
...
-    }
-
-    PyErr_Clear();
-    if (PyErr_WarnEx(
-            PyExc_DeprecationWarning,
-            "Use of the default_value method with arguments is deprecated. "
-            "To set defaults, use set_default_value instead.",
-            1)
-        != 0) {
+    if (!PyArg_ParseTuple(args, "")) {
```
Comment (rahulporuri) [inline traits/ctraits.c:3088]: Why should this function accept `args` now? Should/can we update the function signature and remove this check?

**Review by rahulporuri — APPROVED** [review]
Mostly LGTM with a couple of comments/questions.

File: traits/ctraits.c (hunk 2 above)
Comment (mdickinson) [reply traits/ctraits.c:3088]: True; we could use `METH_NOARGS` here now. Will fix.

**Comment by mdickinson** [comment]
@rahulporuri Thanks for the review! I've updated. If you have bandwidth for another look, that would be much appreciated.

File: traits/ctraits.c (hunk 2 above)
Comment (mdickinson) [reply traits/ctraits.c]: Done in https://github.com/enthought/traits/pull/1632/commits/b7ad8f98f78f87c9367f48d44858286bab6f1167

File: traits/ctraits.c [hunk 3]
```diff
@@ -3094,34 +3094,20 @@ _trait_set_default_value(trait_object *trait, PyObject *args)
...
-|  Get or set the 'default_value_type' and 'default_value' fields
-|  of a CTrait instance. Use of this function for setting the default
-|  value information is deprecated; use set_default_value instead.
+|  Get the 'default_value_type' and 'default_value' fields
+|  of a CTrait instance.
 +----------------------------------------------------------------------------*/
 
 static PyObject *
-_trait_default_value(trait_object *trait, PyObject *args)
+_trait_default_value(trait_object *trait, PyObject *Py_UNUSED(ignored))
```
Comment (rahulporuri) [inline traits/ctraits.c:3102]: Do we need to define a second argument to the `_trait_default_value` function, which is then marked as ignored (IIUC this is done only to silence compiler warnings)?

**Review by rahulporuri — APPROVED** [review]
LGTM with a question

File: traits/ctraits.c (hunk 3 above)
Comment (mdickinson) [reply traits/ctraits.c:3102]: Yes, a second argument is required here. You can see lots of uses of this pattern in the CPython source, too. The function has to be of type `PyCFunction` (docs [here](https://docs.python.org/3/c-api/structures.html)), which is defined as
```
typedef PyObject *(*PyCFunction)(PyObject *, PyObject *);
```

Comment (mdickinson) [reply traits/ctraits.c:3102]: and yes, the `Py_UNUSED` is there to silence compiler warnings.

Comment (rahulporuri) [reply traits/ctraits.c:3102]: Thanks for the reference and explanation. Still LGTM

**Pull request merged**